Change history and System Architectures (#997)
open-contracting/standard: docs/guidance/publish/system_architectures.md

# System architectures

The publication of OCDS data typically requires the design of an architecture that extracts data from live systems, converts it to OCDS format and serves it to users and third party systems.

The design of an architecture depends on several factors:

* Whether existing systems will be modified, or a new system will be built,
* The number and nature of the live systems,
* The technical resources available to the publisher, like storage and processing capabilities, and the availability of technical personnel to maintain systems and modules.

In the ideal scenario:

* Individual releases and records for each contracting process should be available at unique persistent URLs;
* Bulk downloads in JSON, CSV (and, if appropriate, Excel) format should be available covering set periods of contracting;
* Users should be able to easily locate the collections of releases and records they want.

> The following guidance describes some high-level approaches that can be used, with their advantages and disadvantages. These approaches are by no means an extensive or exclusive list, but can be used to inform the design of OCDS publication systems.

Reviewer: ```suggestion
The following guidance describes some high-level approaches that can be used to publish OCDS data, with their advantages and disadvantages. These approaches are by no means an extensive or exclusive list, but can be used to inform the design of OCDS publication systems.
```

### Pull and convert

In this scenario a middleware system sits between live systems and the internet facing API. An automated process periodically pulls data from live systems to the middleware system which performs the conversion to OCDS and maintains a datastore in OCDS format.

![Pull and Convert](../../_static/png/pullAndConvert.png)

The key benefit of this approach is that the middleware system can store a change history, even where this is not maintained in individual source systems.

This approach requires limited modification to live systems, to enable the middleware to pull data. The OCDS Conversion module can merge data from different systems (e.g. when data is extracted from a procurement system and a financial system). To add further source systems, the OCDS conversion element of the middleware needs to be updated to pull data from the new system and convert it to OCDS.

An important decision in the implementation of such a system is how frequently to pull data from the source systems. If data is pulled infrequently there is a risk that the detail of individual changes may be lost.

> An alternative to the pull mechanism is to implement a push mechanism in each source system, triggered by specific events or changes to the data, however approach may require more modifications to the source system.

Reviewer: ```suggestion
An alternative to the pull mechanism is to implement a push mechanism in each source system, triggered by specific events or changes to the data. This approach can mitigate the risk of individual changes being lost, however it may require more modifications to the source system.
```

# System architectures

The publication of OCDS data involves the creation of a conversion process. Like an ETL process, data has to be extracted from one or more sources, converted and either served or stored.

This process needs an adequate architecture to support it. Its design depends on several factors:

* Whether is possible to extend existing systems, or a new system will be built,
* The number and nature of the live systems,
* The technical resources available to the publisher, like storage and processing capabilities. This includes the availability of technical personnel to maintain systems and modules.

Other considerations that affect the design are:

* Individual OCDS documents for each process should be available at unique persistent URLs.
* Bulk downloads in JSON, CSV (and, if appropriate, Excel) formats should be available. These files should be segmented according to one or more criteria, like time periods.
* Users should be able to locate the collections of releases and records they want.

> This guidance describes some design approaches with their advantages and disadvantages. This is not an extensive list, but it can be used to inform the design of the publication system.

Reviewer: ```suggestion
This guidance describes some design approaches with their advantages and disadvantages. This is not an exhaustive list, but it can be used to inform the design of the publication system.
```

## On-demand transformation from live systems

> In this scenario, each source system converts data in OCDS format on demand. Data is not stored but gets converted each time an user o third party invokes the conversion process. This is the easiest path when a single live system manages all contracting processes. But it requires modifications to the system to add an OCDS conversion module.

Reviewer: ```suggestion
In this scenario, each source system converts data in OCDS format on demand. Data is not stored but gets converted each time a user or third party invokes the conversion process. This is the easiest path when a single live system manages all contracting processes. But it requires modifications to the system to add an OCDS conversion module.
```

An API performs data transformation on the fly each time it receives a request.

![Direct Publication 2](../../_static/png/directPublication2.png)

The conversion module produces OCDS releases and/or records wrapped in packages.

This approach does not need extra storage space. But it may be not possible to provide persistent URLs for releases, nor a change history for each process.

> We can think of releases as snapshots of a contracting process in a given time. To be able to rebuild an snapshot for the past depends on how much historical data the source system saves. Also, we need to determine how to identify a snapshot, e.g. if a version id is available in the source system. The Easy releases guidance explains how to achieve a conformant OCDS implementation where it is not possible to provide a full change history.

Reviewer: I think it's better to leave the discussion of releases and records to the other docs pages and just reference them from here.

```suggestion
The easy releases guidance explains how to achieve a conformant OCDS implementation where it is not possible to provide a full change history.
```

Bulk downloads can be provided as part of the API. Live queries may stress the live systems if they need to scan large portions of data.

## Separate OCDS datastore

> In the scenarios that follow, the process converts and stores the data in OCDS format. This has some advantages:

Reviewer: ```suggestion
In the scenarios that follow, a middleware component converts and stores the data in OCDS format. This has some advantages:
```

* It is possible to merge an centralize data from more than one systems in a single datastore.
* It may relieve live systems from expensive queries.
* It may enable the generation of the change history for each contracting process.

> In the other hand, there is a cost of maintaining a separate datastore. In these scenarios, we assume an API provides access to OCDS data.

Reviewer: ```suggestion
On the other hand, there is a cost of maintaining a separate datastore. In these scenarios, we assume an API provides access to OCDS data.
```

Publishers should consider how to store OCDS data. Releases are immutable so can be stored as they are, but records change over time. The process can build records on each API call, or store and update them each time a new release is created. The API must return OCDS data wrapped in a release or record package. Usually there is no need to store wrapped OCDS data, since package data can be generated in real time.

> The change history guidance describes OCDS releases and records and their different components.

Reviewer: Add link

### Pull and convert

> In this scenario a middleware system sits between live systems and the API. An automated process pulls data from live systems to the middleware system. The last performs the conversion to OCDS and maintains a datastore in OCDS format.

Reviewer: ```suggestion
In this scenario an automated process pulls data from live systems to the middleware system. The middleware performs the conversion to OCDS and maintains a datastore in OCDS format.
```

![Pull and Convert](../../_static/png/pullAndConvert.png)

The key benefit of this approach is that the middleware system can store the change history. This is especially good when source systems do not maintain historic data.

> This approach may need changes to live systems to allow the middleware to pull data. The middleware system allows to merge and centralize data in a single place.

Reviewer: ```suggestion
This approach may need some changes to live systems to allow the middleware to pull data. The middleware system merges and centralizes data in a single place.
```

To add more source systems, the OCDS conversion module needs to be updated to pull data from the new system(s).

> An important decision in the implementation is the frequency to pull data. If the frequency is low, there is a risk to lose the detail of individual changes.

Reviewer: ```suggestion
An important decision in the implementation is the frequency to pull data. If the frequency is low, there is a risk of losing the detail of individual changes.
```
